lib/base/tlsstream.cpp

```diff
@@ -103,3 +104,60 @@ void UnbufferedAsioTlsStream::BeforeHandshake(handshake_type type)
 	}
 #endif /* SSL_CTRL_SET_TLSEXT_HOSTNAME */
 }
+
+/**
+ * Forcefully close the connection, typically (details are up to the operating system) using a TCP RST.
+ */
+void AsioTlsStream::ForceDisconnect()
+{
+	if (!lowest_layer().is_open()) {
+		// Already disconnected, nothing to do.
+		return;
+	}
+
+	boost::system::error_code ec;
+
+	// Close the socket. In case the connection wasn't shut down cleanly by GracefulDisconnect(), the operating system
+	// will typically terminate the connection with a TCP RST. Otherwise, this just releases the file descriptor.
+	lowest_layer().close(ec);
+}
+
+/**
+ * Try to cleanly shut down the connection. This involves sending a TLS close_notify shutdown alert and terminating the
+ * underlying TCP connection. Sending these additional messages can block, hence the method takes a yield context and
+ * internally implements a timeout of 10 seconds for the operation after which the connection is forcefully terminated
+ * using ForceDisconnect().
+ *
+ * @param strand Asio strand used for other operations on this connection.
+ * @param yc Yield context for Asio coroutines
+ */
+void AsioTlsStream::GracefulDisconnect(boost::asio::io_context::strand& strand, boost::asio::yield_context& yc)
+{
+	if (!lowest_layer().is_open()) {
+		// Already disconnected, nothing to do.
+		return;
+	}
+
+	boost::system::error_code ec;
+
+	Timeout::Ptr shutdownTimeout(new Timeout(strand.context(), strand, boost::posix_time::seconds(10),
+		[this, keepAlive = AsioTlsStream::Ptr(this)](boost::asio::yield_context yc) {
+			// Forcefully terminate the connection if async_shutdown() blocked more than 10 seconds.
+			ForceDisconnect();
```

Comment:
So you made the whole a85c188fed19d4751645d8820fab5985d923aea5 for `keepAlive`. ForceDisconnect() is also called below, so `keepAlive` is only for going out of scope w/o cancelling. So wouldn't it be nice to get a timeout implicitly cancelled while going out of scope?

Reply:
Inside GracefulDisconnect() itself you're sure `this` is kept alive by its caller, right? So it's only the Timeout callback which _could theoretically be called_* outside GracefulDisconnect() scope and hence needs `keepAlive`.

*) I've fixed this:

* #10254

Reply:
First of all, if someone requests a review from you, your response to that **should not** be a new PR with **+100** code changes. Requesting a review doesn't mean replacing someone's implementation with your own ones, that's just bad and you just stacking blocking PRs on one another. I don't mind if you want to clean up the `timeout` class, but your newly opened PRs have nothing to do with this one. #10210 is stalled because it's still waiting for this PR to be merged, bzw. when it should have been merged already last week, and now you're forcing this one to be blocked by some random PRs. So, please **STOP** doing that! Opening a new PR means spending more time reviewing and testing on something that isn't relevant now.

Reply:
> > First of all, if someone requests a review from you, your response to that **should not** be a new PR with **+100** code changes.
> 
> Fact check: #10254, **which Julian seems to like** btw. IIRC (and even expressed sympathy to #10250 in [#10005 (comment)](https://github.com/Icinga/icinga2/pull/10005#discussion_r1856766497))

Just because I think something is a good idea in general doesn't mean that this should block this PR.

> > Requesting a review doesn't mean replacing someone's implementation with your own ones, that's just bad and you just stacking blocking PRs on one another.
> 
> In general, yes. I "also" don't like if someone tells me my implementation is bad, like in #10117 😈, but I don't do that here. In fact, **4/5 commits here are almost perfectly fine!**

So what's the problem then with this PR? Isn't `Shared<AsioTlsStream>::Ptr` `AsioTlsStream::Ptr` essentially the same thing, just that one allows you to safely (as long as you own a shared pointer) get from `AsioTlsStream` to the corresponding `Ptr`?

> > #10210 is stalled because it's still waiting for this PR to be merged, bzw. when it should have been merged already last week,
> 
> We already have packages, haven't we?

This is not about any packages, it is about continuing development in general.

Reply:
> Just because I think something is a good idea in general doesn't mean that this should block this PR.

For a "good idea", which is also relatively small btw. (if you count just the resulting Timeout class), you look surprisingly hesitant to me. Is that "good idea" that far from an "✅ #10254 LGTM", especially given the high unit tests coverage? _Even I_ have made several compromises (standing in diametral contrast to my professional opinion), e.g. in #10117 – why shouldn't you too?

> So what's the problem then with this PR? Isn't `Shared<AsioTlsStream>::Ptr` `AsioTlsStream::Ptr` essentially the same thing, just that one allows you to safely (as long as you own a shared pointer) get from `AsioTlsStream` to the corresponding `Ptr`?

But this is not how it should work. In a perfect world, the less stuff needs to drag _itself_ around via intrusive pointers or in general rely on the fact that the caller holds it via intrusive pointers, the better. Especially there should be well defined RAII scopes, as in #10254, for this. In opposite to a refactoring to be able to drag stuff around via intrusive pointers.

> This is not about any packages, it is about continuing development in general.

Honestly speaking, I'm ~a bit~ puzzled. So... you simply want to get this done? Let's just merge #10254 then. (But instead you have written https://github.com/Icinga/icinga2/pull/10005#discussion_r1861780232?) I'd even adjust #10005 afterwards at your option.

I mean, it's not that I always behave like with #10254 here. Neither I fold my arms and abstractly claim something must work more beautiful. I've presented a specific constructive solution you even consider a "good idea".